### ISSUE TITLE
EHCache not working in native mode with docker builds

### ISSUE THREAD

**Issue by antonwiens** [quarkiverse/quarkus-cxf#456]
As mentioned in the related issue, i am having problems getting ehcache to run with quarkus 2.10.0.CR1 and quarkus-cxf 1.1.0.

I do not exclude the ehcache dependency or anything related to it.
Do you have any idea why i get this error message?

```
Caused by: org.apache.wss4j.common.ext.WSSecurityException: org.ehcache.xml.exceptions.XmlConfigurationException: Error parsing XML configuration at resource:/cxf-ehcache.xml
Original Exception was org.apache.cxf.ws.security.tokenstore.TokenStoreException: org.ehcache.xml.exceptions.XmlConfigurationException: Error parsing XML configuration at resource:/cxf-ehcache.xml
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.getCallback(WSS4JInInterceptor.java:630)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.handleMessageInternal(WSS4JInInterceptor.java:271)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.handleMessage(WSS4JInInterceptor.java:213)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.handleMessage(WSS4JInInterceptor.java:93)
        at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:307)
        at org.apache.cxf.endpoint.ClientImpl.onMessage(ClientImpl.java:829)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponseInternal(HTTPConduit.java:1701)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponse(HTTPConduit.java:1570)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1371)
        at org.apache.cxf.metrics.interceptors.CountingOutputStream.close(CountingOutputStream.java:47)
        at org.apache.cxf.transport.AbstractConduit.close(AbstractConduit.java:56)
        at org.apache.cxf.transport.http.HTTPConduit.close(HTTPConduit.java:671)
        at org.apache.cxf.interceptor.MessageSenderInterceptor$MessageSenderEndingInterceptor.handleMessage(MessageSenderInterceptor.java:63)
        at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:307)
        at org.apache.cxf.endpoint.ClientImpl.doInvoke(ClientImpl.java:528)
        at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:439)
        at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:354)
        at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:312)
        at org.apache.cxf.frontend.ClientProxy.invokeSync(ClientProxy.java:96)
        at org.apache.cxf.jaxws.JaxWsClientProxy.invoke(JaxWsClientProxy.java:140)
        ... 43 more
Caused by: org.apache.cxf.ws.security.tokenstore.TokenStoreException: org.ehcache.xml.exceptions.XmlConfigurationException: Error parsing XML configuration at resource:/cxf-ehcache.xml
        at org.apache.cxf.ws.security.tokenstore.EHCacheTokenStore.<init>(EHCacheTokenStore.java:82)
        at org.apache.cxf.ws.security.tokenstore.EHCacheTokenStoreFactory.newTokenStore(EHCacheTokenStoreFactory.java:45)
        at org.apache.cxf.ws.security.tokenstore.TokenStoreUtils.getTokenStore(TokenStoreUtils.java:58)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.getCallback(WSS4JInInterceptor.java:688)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.getCallback(WSS4JInInterceptor.java:628)
        ... 62 more
Caused by: org.ehcache.xml.exceptions.XmlConfigurationException: Error parsing XML configuration at resource:/cxf-ehcache.xml
        at org.ehcache.xml.XmlConfiguration.<init>(XmlConfiguration.java:124)
        at org.apache.cxf.ws.security.tokenstore.EHCacheTokenStore.<init>(EHCacheTokenStore.java:69)
        ... 66 more
Caused by: java.nio.file.NoSuchFileException: /project/lib/org.ehcache.ehcache-3.9.6.jar
        at sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
        at sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:149)
        at sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
        at java.nio.file.Files.readAttributes(Files.java:1764)
        at java.util.zip.ZipFile$Source.get(ZipFile.java:1259)
        at java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:733)
        at java.util.zip.ZipFile$CleanableResource.get(ZipFile.java:850)
        at java.util.zip.ZipFile.<init>(ZipFile.java:248)
        at java.util.zip.ZipFile.<init>(ZipFile.java:177)
        at java.util.jar.JarFile.<init>(JarFile.java:350)
        at sun.net.www.protocol.jar.URLJarFile.<init>(URLJarFile.java:103)
        at sun.net.www.protocol.jar.URLJarFile.getJarFile(URLJarFile.java:72)
        at sun.net.www.protocol.jar.JarFileFactory.get(JarFileFactory.java:99)
        at sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:125)
        at sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:155)
        at java.net.URL.openStream(URL.java:1165)
        at org.ehcache.xml.ConfigurationParser.<init>(ConfigurationParser.java:158)
        at org.ehcache.xml.XmlConfiguration.<init>(XmlConfiguration.java:114)
        ... 67 more

```

_Originally posted by @antonwiens in https://github.com/quarkiverse/quarkus-cxf/issues/443#issuecomment-1156385024_

**Comment by antonwiens** [comment]
I am using gradle and can see the following problem:

org.apache.cxf:cxf-rt-ws-security:3.5.2 is pulled in by quarkus-cxf-rt-ws-security.
Which is pulling ehcache 3.9.6 over org.apache.wss4j:wss4j-ws-security-dom:2.4.1
and 3.9.3 over rg.apache.cxf:cxf-rt-security-saml:3.5.2.

Gradle is taking 3.9.6 there to replace 3.9.3.

I will try forcing either version to fix this problem.

**Comment by antonwiens** [comment]
This is the result of trying to force ehcache 3.9.3:

```
Caused by: org.apache.wss4j.common.ext.WSSecurityException: org.ehcache.xml.exceptions.XmlConfigurationException: Error parsing XML configuration at resource:/cxf-ehcache.xml
Original Exception was org.apache.cxf.ws.security.tokenstore.TokenStoreException: org.ehcache.xml.exceptions.XmlConfigurationException: Error parsing XML configuration at resource:/cxf-ehcache.xml
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.getCallback(WSS4JInInterceptor.java:630)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.handleMessageInternal(WSS4JInInterceptor.java:271)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.handleMessage(WSS4JInInterceptor.java:213)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.handleMessage(WSS4JInInterceptor.java:93)
        at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:307)
        at org.apache.cxf.endpoint.ClientImpl.onMessage(ClientImpl.java:829)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponseInternal(HTTPConduit.java:1701)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleResponse(HTTPConduit.java:1570)
        at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1371)
        at org.apache.cxf.metrics.interceptors.CountingOutputStream.close(CountingOutputStream.java:47)
        at org.apache.cxf.transport.AbstractConduit.close(AbstractConduit.java:56)
        at org.apache.cxf.transport.http.HTTPConduit.close(HTTPConduit.java:671)
        at org.apache.cxf.interceptor.MessageSenderInterceptor$MessageSenderEndingInterceptor.handleMessage(MessageSenderInterceptor.java:63)
        at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:307)
        at org.apache.cxf.endpoint.ClientImpl.doInvoke(ClientImpl.java:528)
        at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:439)
        at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:354)
        at org.apache.cxf.endpoint.ClientImpl.invoke(ClientImpl.java:312)
        at org.apache.cxf.frontend.ClientProxy.invokeSync(ClientProxy.java:96)
        at org.apache.cxf.jaxws.JaxWsClientProxy.invoke(JaxWsClientProxy.java:140)
        ... 43 more
Caused by: org.apache.cxf.ws.security.tokenstore.TokenStoreException: org.ehcache.xml.exceptions.XmlConfigurationException: Error parsing XML configuration at resource:/cxf-ehcache.xml
        at org.apache.cxf.ws.security.tokenstore.EHCacheTokenStore.<init>(EHCacheTokenStore.java:82)
        at org.apache.cxf.ws.security.tokenstore.EHCacheTokenStoreFactory.newTokenStore(EHCacheTokenStoreFactory.java:45)
        at org.apache.cxf.ws.security.tokenstore.TokenStoreUtils.getTokenStore(TokenStoreUtils.java:58)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.getCallback(WSS4JInInterceptor.java:688)
        at org.apache.cxf.ws.security.wss4j.WSS4JInInterceptor.getCallback(WSS4JInInterceptor.java:628)
        ... 62 more
Caused by: org.ehcache.xml.exceptions.XmlConfigurationException: Error parsing XML configuration at resource:/cxf-ehcache.xml
        at org.ehcache.xml.XmlConfiguration.<init>(XmlConfiguration.java:124)
        at org.apache.cxf.ws.security.tokenstore.EHCacheTokenStore.<init>(EHCacheTokenStore.java:69)
        ... 66 more
Caused by: java.nio.file.NoSuchFileException: /project/lib/org.ehcache.ehcache-3.9.3.jar
        at sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
        at sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:149)
        at sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
        at java.nio.file.Files.readAttributes(Files.java:1764)
        at java.util.zip.ZipFile$Source.get(ZipFile.java:1259)
        at java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:733)
        at java.util.zip.ZipFile$CleanableResource.get(ZipFile.java:850)
        at java.util.zip.ZipFile.<init>(ZipFile.java:248)
        at java.util.zip.ZipFile.<init>(ZipFile.java:177)
        at java.util.jar.JarFile.<init>(JarFile.java:350)
        at sun.net.www.protocol.jar.URLJarFile.<init>(URLJarFile.java:103)
        at sun.net.www.protocol.jar.URLJarFile.getJarFile(URLJarFile.java:72)
        at sun.net.www.protocol.jar.JarFileFactory.get(JarFileFactory.java:99)
        at sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:125)
        at sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:155)
        at java.net.URL.openStream(URL.java:1165)
        at org.ehcache.xml.ConfigurationParser.<init>(ConfigurationParser.java:149)
        at org.ehcache.xml.XmlConfiguration.<init>(XmlConfiguration.java:114)
        ... 67 more

```

**Comment by shumonsharif** [comment]
Strange indeed ... I would love to help you, but this does not appear in any way to be an issue with the extension.  Please feel free to put together a simple reproducer project if you'd like for me to try to take a look?

**Comment by antonwiens** [comment]
The problem seems to be related to org.ehcache.xml.XmlConfiguration#CORE_SCHEMA_URL. 
It gets initialized like this:

```
    public static final URL CORE_SCHEMA_URL = XmlConfiguration.class.getResource("/ehcache-core.xsd");
```

And it seems like it is loaded in org/ehcache/xml/ConfigurationParser.java:158.

But i can`t find where my problem is coming from...

When i have time, i will try to create a reproducer.

**Comment by shumonsharif** [comment]
The attached [gradle-ws-security.zip](https://github.com/quarkiverse/quarkus-cxf/files/8957496/gradle-ws-security.zip) gradle project executes without any issues for me.  I'm still thinking there's something off in your setup, and likely the issue has nothing to do with the actual loading of the xsd in XmlConfiguration.

Can you try run this project; once you run the native executable, try something like:

```bash
curl -X POST "http://localhost:8080/soap/greeting-service" \
 -H 'Content-Type: text/xml' \
 -H 'SOAPAction:' \
 -d '<soapenv:Envelope xmlns:acme="http://acme.org/" xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
   <soapenv:Header><wsse:Security xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd"><wsse:UsernameToken><wsse:Username>joe</wsse:Username><wsse:Password Type="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-username-token-profile-1.0#PasswordText">wss4j</wsse:Password></wsse:UsernameToken></wsse:Security></soapenv:Header>
   <soapenv:Body>
      <acme:hello>
         <arg0>Shumon</arg0>
      </acme:hello>
   </soapenv:Body>
</soapenv:Envelope>'
```

**Comment by shumonsharif** [comment]
You're right, I stand corrected - the issue appears to be specific to the docker images.  Things are fine when running the native binary without docker.  Will dig in and keep you posted.

**Comment by shumonsharif** [comment]
Hey @antonwiens So I dug in a little, and it may take a while for me to identify and address the root cause of this issue.  I'm posting a temporary workaround below until I can figure out what's happening.

```java
import io.quarkus.arc.Unremovable;
import io.quarkus.runtime.StartupEvent;
import org.apache.cxf.Bus;
import org.apache.cxf.BusFactory;
import org.apache.cxf.ws.security.tokenstore.MemoryTokenStore;
import org.apache.cxf.ws.security.tokenstore.TokenStore;

import javax.enterprise.event.Observes;

@Unremovable
public class TokenStoreConfigurer {

    void onStart(@Observes StartupEvent ev) {
        final Bus bus = BusFactory.getThreadDefaultBus();
        bus.setProperty(TokenStore.class.getName(), new MemoryTokenStore());
    }

}
```

**Comment by antonwiens** [comment]
> Hey @antonwiens So I dug in a little, and it may take a while for me to identify and address the root cause of this issue. I'm posting a temporary workaround below until I can figure out what's happening.
> 
> ```java
> import io.quarkus.arc.Unremovable;
> import io.quarkus.runtime.StartupEvent;
> import org.apache.cxf.Bus;
> import org.apache.cxf.BusFactory;
> import org.apache.cxf.ws.security.tokenstore.MemoryTokenStore;
> import org.apache.cxf.ws.security.tokenstore.TokenStore;
> 
> import javax.enterprise.event.Observes;
> 
> @Unremovable
> public class TokenStoreConfigurer {
> 
>     void onStart(@Observes StartupEvent ev) {
>         final Bus bus = BusFactory.getThreadDefaultBus();
>         bus.setProperty(TokenStore.class.getName(), new MemoryTokenStore());
>     }
> 
> }
> ```

Thanks for looking into it and the workaround. If i understand correctly the workaround disables EHCache for the TokenStore for the time being?

**Comment by shumonsharif** [comment]
> Thanks for looking into it and the workaround. If i understand correctly the workaround disables EHCache for the TokenStore for the time being?

That's correct.  Keep in mind there's a difference in the token TTL, which may have some ramifications if you have a high throughput application.  You can tune the TTL by setting the value directly on the `MemoryTokenStore` instance.  

From the `EHCacheTokenStore` docs:  An in-memory EHCache implementation of the TokenStore interface. The default TTL is 60 minutes and the max TTL is 12 hours.

From the `MemoryTokenStore` docs:  A simple HashMap-based TokenStore. The default TTL is 5 minutes and the max TTL is 1 hour.

Just wanted to note: this page lists a whole bunch of properties you can use to configure WS-Security:
https://cxf.apache.org/docs/ws-securitypolicy.html
I believe you had asked about disabling ehcache in the related issue.  You should be able to use the properties defined on this page to do so.

**Comment by antonwiens** [comment]
> > Thanks for looking into it and the workaround. If i understand correctly the workaround disables EHCache for the TokenStore for the time being?
> 
> That's correct. Keep in mind there's a difference in the token TTL, which may have some ramifications if you have a high throughput application. You can tune the TTL by setting the value directly on the `MemoryTokenStore` instance.
> 
> From the `EHCacheTokenStore` docs: An in-memory EHCache implementation of the TokenStore interface. The default TTL is 60 minutes and the max TTL is 12 hours.
> 
> From the `MemoryTokenStore` docs: A simple HashMap-based TokenStore. The default TTL is 5 minutes and the max TTL is 1 hour.
> 
> Just wanted to note: this page lists a whole bunch of properties you can use to configure WS-Security: https://cxf.apache.org/docs/ws-securitypolicy.html I believe you had asked about disabling ehcache in the related issue. You should be able to use the properties defined on this page to do so.

Thanks for all the hints.

**Comment by ppalaga** [comment]
This is reproducible with the test in https://github.com/quarkiverse/quarkus-cxf/pull/503/commits/58c58cb450991e27d3a4d8a6bf6d3f16e95b12e0
and fixed by upgrading to Ehcache 3.10.0 and moving the init of some classes to runtime - see https://github.com/quarkiverse/quarkus-cxf/pull/503/commits/cac66c138084a704264d47aee6c15cc40a1b808a
